mono/metadata/superluminal.c

```diff
@@ -0,0 +1,330 @@
+#include <mono/metadata/tokentype.h>
+#include <mono/metadata/unity-utils.h>
+
+#include <string.h>
+
+#define WIN32_LEAN_AND_MEAN
+#include "Evntcons.h"
+#include "evntprov.h"
+#include "evntrace.h"
+#include "cguid.h"
+#include "windows.h"
+
+DECLSPEC_NOINLINE __inline VOID __stdcall Private_EventControlCallback (_In_ LPCGUID SourceId, _In_ ULONG ControlCode, _In_ UCHAR Level, _In_ ULONGLONG MatchAnyKeyword, _In_ ULONGLONG MatchAllKeyword, _In_opt_ PEVENT_FILTER_DESCRIPTOR FilterData, _Inout_opt_ PVOID CallbackContext);
+#define MCGEN_PRIVATE_ENABLE_CALLBACK_V2 Private_EventControlCallback
+#include "superluminal-CLR-ETW-Generated.h"
+
+#define MAX_NUM_OFFSETS 7000
+#define ENABLE_VERBOSE_LOGGING 0
+
+static gboolean is_initialized = FALSE;
+
+#if ENABLE_VERBOSE_LOGGING
+	static void
+	sLogInternal(const char* inFormat, ...)
+	{
+		va_list args;
+		va_start (args, inFormat);
+
+		char buffer[1024] = { 0 };
+		vsprintf_s (buffer, 1024, inFormat, args);
+
+		va_end(args);
+
+		OutputDebugStringA (buffer);
+	}
+
+	#define SUPERLUMINAL_LOG_ARGS(format, ...) sLogInternal("SUPERLUMINAL: "format"\n", __VA_ARGS__)
+	#define SUPERLUMINAL_LOG(str) sLogInternal("SUPERLUMINAL: %s\n", str)
+#else
+	#define SUPERLUMINAL_LOG_ARGS(format, ...)
+	#define SUPERLUMINAL_LOG(str)
+#endif
+
+static void
+image_event (MonoImage *image, gboolean is_rundown, gboolean is_unload)
+{
+	if (!MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_Context.IsEnabled && !MICROSOFT_WINDOWS_DOTNETRUNTIME_RUNDOWN_PROVIDER_Context.IsEnabled)	{
+		SUPERLUMINAL_LOG ("Providers not enabled, skipping image_event");
+		return;
+	}
+
+	const char *pdb_path = NULL;
+	guint8 pe_guid[16] = {0};
+	gint32 pe_age = 0;
+	gint32 pe_timestamp = 0;
+
+	// We only emit PDB info for loads *and* if the image is not a dynamic image (i.e. containing dynamic methods). This is becuase
+	// dynamic images do not represent an actual on-disk image and so don't have any PDB info (calling mono_ppdb_get_signature will lead to a crash)
+	if (!is_unload && !mono_image_is_dynamic (image))
+		mono_ppdb_get_signature (image, &pdb_path, pe_guid, &pe_age, &pe_timestamp);
+
+	gunichar2 *image_path_utf16 = u8to16 (mono_image_get_filename (image));
+	gunichar2 *pdb_path_utf16 = pdb_path != NULL ? u8to16 (pdb_path) : NULL;
+
+	MonoAssembly *assembly = mono_image_get_assembly (image);
+
+	if (is_rundown) {
+		EventWriteModuleDCEnd_V2 ((uint64_t)image, (uint64_t)assembly, 0, 0, image_path_utf16, L"", 0, (GUID *)pe_guid, pe_age, pdb_path_utf16 == NULL ? L"" : pdb_path_utf16, &GUID_NULL, 0, L"");
+	} else if (is_unload) {
+		EventWriteModuleUnload_V2 ((uint64_t)image, (uint64_t)assembly, 0, 0, image_path_utf16, L"", 0, (GUID *)pe_guid, pe_age, pdb_path_utf16 == NULL ? L"" : pdb_path_utf16, &GUID_NULL, 0, L"");
+	} else {
+		EventWriteModuleLoad_V2 ((uint64_t)image, (uint64_t)assembly, 0, 0, image_path_utf16, L"", 0, (GUID *)pe_guid, pe_age, pdb_path_utf16 == NULL ? L"" : pdb_path_utf16, &GUID_NULL, 0, L"");
```

Comment:
@joshuap re missing module name: my suspicion is that either `image_path_utf16` empty (because Unity loads from disk),  or the parameter that follows (the `L""`) supposed to be module name. I'm not 100% on this, though, but that's the impression I get when comparing it with CoreCLR source code here: https://github.com/Unity-Technologies/runtime/blob/unity-main/src/coreclr/vm/eventtrace.cpp#L6381

Reply:
Some more observations from looking at the data:

1. We are not emitting Assembly load events, so assembly we're passing in refers to "Unknown" (compare it with Unity.Licensing.Client.exe which runs on CoreCLR:
![image](https://user-images.githubusercontent.com/1092489/158489839-35b773e7-7a54-40bb-b573-b09266d748e3.png)

2. I'm seeing some really weird strings in ModuleILPath: 
![image](https://user-images.githubusercontent.com/1092489/158489952-1990a9db-d5e3-498d-8a57-b8b9ea46a07e.png)

Reply:
I also wonder what's the significance of ignoring "unload" events. Could this cause bad stuff to happen during domain unloads if the new domain happens to reuse the same memory addresses?

Reply:
hey @TautvydasZilys @joshpeterson, some responses to your questions:

1. Can you explain where you're missing the module name? `image_path_utf16` is the path to the image and it corresponds to `ModuleILPath` in the ETW event. The `L""` parameter that follows corresponds to `ModuleNativePath` in the ETW event, which is used to communicate the path to the NGEN/ReadyToRun image and is only relevant for AOT scenarios, which I don't think is applicable for Unity?
2. We could definitely add support for assembly load/unload. I'm not sure what they'd be used for, but there is a corresponding event in the mono profiler interface, so it shouldn't be too hard to add.
3. The weird ModuleILPaths are emitted for in-memory images that don't have a corresponding on-disk image. It corresponds to a call to `mono_image_open_from_data_internal`.
4. Assuming you're talking about MethodUnload events, yes, those are currently missing. The reason why they're not there is because the mono profiler interface has no corresponding callback when JIT methods are destroyed. When we were working on our original PR (https://github.com/Unity-Technologies/mono/pull/1369), we actually added a `jit_unload` event to the mono profiler interface, but that turned out to be quite complicated, and it was unclear whether that actually covered all cases. So in the end the PR was made without that event in. If we want to emit MethodUnload events, the mono profiler interface will need to be updated with a callback for that. I'm not sure if it's actually a problem if these are missing. We've done a lot of testing on our end with entering/leaving playmode and such (which results in domain reloads & unloads), and Superluminal at least deals with it just fine (including when memory addresses are reused), but it's hard to say how other tools like WPA and PerfView might deal with that.

Reply:
> The weird ModuleILPaths are emitted for in-memory images that don't have a corresponding on-disk image. It corresponds to a call to mono_image_open_from_data_internal.

Yeah, I just noticed this as well. They seem to come from Mono loading ppdb files as "images". They load with the `metadata_only` flag set to true. Could we skip images with that flag set? That would eliminate them from showing up. I don't think this would have any adverse impacts, because there should never be any code we care to profile in them, right?

Reply:
I'm pushed a new change to implement this check to skip metadata only images.

Reply:
> Can you explain where you're missing the module name? 

I'm looking at raw events in WPA and the names are missing there: 
![image](https://user-images.githubusercontent.com/1092489/158667921-9b8e84de-f3c8-4f08-9d14-7e737a4dfa81.png)

> image_path_utf16 is the path to the image and it corresponds to ModuleILPath in the ETW event. The L"" parameter that follows corresponds to ModuleNativePath in the ETW event, which is used to communicate the path to the NGEN/ReadyToRun image and is only relevant for AOT scenarios, which I don't think is applicable for Unity?

You are right, after looking at it some more I suspect that it's due to assembly loads missing.

I don't think that it blocks this PR from landing as is, though. It would be nice to have those image names for completeness, but this is already is so game changing.

Reply:
> I'm pushed a new change to implement this check to skip metadata only images.

Cool, LGTM :-)

> You are right, after looking at it some more I suspect that it's due to assembly loads missing.
> 
> I don't think that it blocks this PR from landing as is, though. It would be nice to have those image names for completeness, but this is already is so game changing.

Can you share the column configuration (or a screenshot) you used to get that output in WPA? I can take a stab at adding the assembly loads in our branch, then @joshpeterson could possibly merge those changes into this version (if it fixes this issue). I don't think it should be too much work at first glance.

Reply:
Sure, there are two giveaways:

1. Under `System Activity` -> `Images` view, managed assemblies are lacking names:

![image](https://user-images.githubusercontent.com/1092489/158710574-0dd6616b-ed91-43ab-9800-0a1606a9e693.png)

2. In any stacktrace view, any managed symbol is also missing the assembly name:
![image](https://user-images.githubusercontent.com/1092489/158710842-005b3715-015c-413a-8042-057040f4b63a.png)

Reply:
Thanks! Will have a look.

Reply:
@joshpeterson @TautvydasZilys I've spent a few hours the morning looking into the missing module thing in WPA and I'm stumped :-) I've added support for writing AssemblyLoad, AssemblyUnload, AppDomainLoad and AppDomainUnload events and with all of that WPA is still not displaying module names in either the callstack or the image list.

At this point I'm really not sure what WPA is missing to be able to display the module names. I've committed the changes to a separate branch [here](https://github.com/SuperluminalHQ/mono/commit/e90852fc4d7cdb76e4388e6e9d51735568ed2e38) in case you want to have a look at it, but I wouldn't recommend integrating it into this PR at this point.

These are the events I'm sending now:
![image](https://user-images.githubusercontent.com/8319699/158772117-9ea1ffff-1e7a-4857-97e7-4ef14786f024.png)

Example of the AssemblyLoad events:
![image](https://user-images.githubusercontent.com/8319699/158772038-00156e50-e182-428e-afbf-068c070b0b17.png)

Reply:
Thanks for digging into this @rovarma! Let's wait to hear @TautvydasZilys opinion, but I suspect that we will merge this PR as-is for the time being.

Reply:
Let's land it for now. We can revisit if this causes pain in the future.